File: CourseProjects/CourseProjects/JavaEclipseProjects/UnitTestingExamples/src/demos/tdd/mockito/basic/ShopTest.java
package demos.tdd.mockito.basic;

import org.junit.Before;
import org.junit.Test;

import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

import static org.mockito.Mockito.*;

public class ShopTest {
	@Before
	public void start() {
		pricingEngine = mock(PricingEngine.class);
		paymentEngine = mock(PaymentEngine.class);
		stockCheckEngine = mock(StockCheckEngine.class);
	}
	@Test
	public void makePurchaseWorksForValidQuantity() {
		when(stockCheckEngine.check("AB12")).thenReturn(20);
		when(pricingEngine.price("AB12",19)).thenReturn(27.30);
		when(paymentEngine.authorize("010203XYZ",27.30)).thenReturn(true);
		
		Shop s = new Shop(pricingEngine,stockCheckEngine,paymentEngine);
		assertTrue(s.makePurchase("AB12", 19, "010203XYZ"));
		
		verify(stockCheckEngine).check("AB12");
		verify(pricingEngine).price("AB12",19);
		verify(paymentEngine).authorize("010203XYZ",27.30);
	}
	@Test
	public void makePurchaseWorksForInvalidQuantity() {
		when(stockCheckEngine.check("AB12")).thenReturn(20);

		Shop s = new Shop(pricingEngine,stockCheckEngine,paymentEngine);
		assertFalse(s.makePurchase("AB12", 21, "010203XYZ"));

		verify(stockCheckEngine).check("AB12");
	}
	private PricingEngine pricingEngine;
	private PaymentEngine paymentEngine;
	private StockCheckEngine stockCheckEngine;
}

File: CourseProjects/CourseProjects/VisualStudioProjects/Agile/MockingWithMoq/BasicMocking/ShopTest.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using Moq;

namespace BasicMocking {
    [TestClass]
    public class ShopTest {
        [TestInitialize]
        public void start() {
            pricingMock = new Mock<IPricingEngine>();
            paymentMock = new Mock<IPaymentEngine>();
            stockCheckMock = new Mock<IStockCheckEngine>();
        }
        [TestMethod]
        public void MakePurchaseWorksForValidQuantity() {
            stockCheckMock.Setup(m => m.Check("AB12")).Returns(20).Verifiable();
            pricingMock.Setup(m => m.Price("AB12", 19)).Returns(27.30).Verifiable();
            paymentMock.Setup(m => m.Authorize("010203XYZ", 27.30)).Returns(true).Verifiable();

            Shop s = new Shop(pricingMock.Object, stockCheckMock.Object, paymentMock.Object);
            Assert.IsTrue(s.makePurchase("AB12", 19, "010203XYZ"));

            stockCheckMock.Verify();
            pricingMock.Verify();
            paymentMock.Verify();
        }
        [TestMethod]
        public void MakePurchaseWorksForInvalidQuantity() {
            stockCheckMock.Setup(m => m.Check("AB12")).Returns(20).Verifiable();

            Shop s = new Shop(pricingMock.Object, stockCheckMock.Object, paymentMock.Object);
            Assert.IsFalse(s.makePurchase("AB12", 21, "010203XYZ"));

            stockCheckMock.Verify();
        }
        private Mock<IPricingEngine> pricingMock;
        private Mock<IPaymentEngine> paymentMock;
        private Mock<IStockCheckEngine> stockCheckMock;
    }
}
